
<!DOCTYPE html>
<html>
<head>
    <title>L1000CDS²</title>
	<meta charset='UTF-8'>
	i
    <link rel="shortcut icon" href="images/sigine.ico" type="image/x-icon" />
	<meta content='width=device-width, initial-scale=1' name='viewport'/>
	<link rel='stylesheet' type='text/css' href='css/main.css'/>
	<link rel='stylesheet' type='text/css' href='css/open-sans.css'/>
</head>
<body>
    <div id='sidebar'>
        <ul>
            <li><a href='#summary'>Summary</a></li>
            <li><a href='#L1000Data'>Description of the L1000 Data</a></li>
            <li><a href='#download'>Signatures Available for Download</a></li>
            <li><a href='#tutorial'>Tutorial</a></li>
            <ul id='sideBarSub'>
                <li><a href="#input" class="subtitle"> Input </a></li>
                <li><a href='#search'>Search</a></li>
                <li><a href='#example'>Examples and Signatures</a></li>
                <li><a href='#config'>Configuration</a></li>
                <li><a href='#metadata'>Metadata</a></li>
                <li><a href='#recent'>Recent Searches</a></li>
                <li><a href="#result" class="subtitle"> Result</a></li>
                <li><a href='#table'>Table View</a></li>
                <li><a href='#header'>Header</a></li>
                 <li><a href='#combinations'>Table of combinations</a></li>
                 <li><a href='#substructures'>Substructure EA</a></li>
                <!-- <li><a href='#share'>Share</a></li> -->
            </ul>
            <li><a href='#api'>API</a></li>
            <ul id='sideBarSub'>
                <li><a href="#geneSet"> Gene-set Search</a></li>
                <li><a href="#cosine"> Cosine Distance Search</a></li>
                <li><a href="#python"> Python 3 Example</a></li>
            </ul>
            <li><a href='#contact'>Contact</a></li>
        </ul>
    </div>
    <div id='content'>
        <h2 id='title'>
            <img id='logo' src='images/sigine.png'/>
            <span>L1000CDS²</span>
        </h2>
        <h4 id='subtitle'>An <span class="highlight lightning-fast">ultra-fast</span> LINCS <span class="highlight">L1000 </span><span class="highlight">C</span>haracteristic <span class="highlight">D</span>irection <span class="highlight">S</span>ignature<span class="highlight"> S</span>earch Engine</h4>
        <section id='summary'>
            <h3 class='first'>Summary</h3>
            <p>L1000CDS² is a LINCS L1000 characteristic direction signature search engine. It enables users to find consensus L1000 small molecule signatures that match user input signatures. The underlying dataset for the search engine is a portion of the <a target="_blank" href="http://www.lincscloud.org/">LINCS L1000</a> small molecule expression profiles generated at the Broad Institute by the <a target="_blank" href="http://lincscloud.org/about/cmap-team/">Connectivity Map team</a>. The differentially expressed (DE) genes of these profiles were calculated using the <a target="_blank" href="http://maayanlab.net/CD/">characteristic direction</a> method. Depending on the user’s input, L1000CDS² uses either a gene-set method or cosine distance method to compare the input signatures to the L1000 signatures to perform the search. When up/down gene lists are submitted to L1000CDS², the search engine compares the input gene lists to the DE genes computed from the LINCS L1000 data and descriptive information of the top 50 matched signatures is returned. When a signature is submitted to L1000CDS² in the format of "gene symbol, expression value", the search engine calculates a cosine distance between the input signature and every characteristic direction signature in the underlying dataset, and the top 50 signatures of either the largest (reverse mode) or the smallest (mimic mode) cosine distances are returned. L1000CDS² leverages the efficiency of matrix operations to perform the search. The search finishes a query against more than 20,000 signatures in less than a  decisecond using the gene-set method or less than 4 seconds using the cosine distance method. The L1000CDS2 application is developed by the <a href='http://icahn.mssm.edu/research/labs/maayan-laboratory' target='_blank'>Ma'ayan Laboratory</a> at the <a href='http://icahn.mssm.edu/' target='_blank'>Icahn School of Medicine at Mount Sinai</a> for the <a href='http://www.lincs-dcic.org/' target='_blank'>BD2K-LINCS-DCIC</a> and the <a href='http://commonfund.nih.gov/idg/fundedresearch' target='_blank'>KMC-IDG </a>NIH funded projects.</p>
        </section>
        <section id='L1000Data'>
            <h3>Description of the L1000 Data</h3>
            <p>The L1000 mRNA gene-expression dataset is generated as part of the <a href="http://www.lincsproject.org/">Library of Integrated Network-based Cellular Signatures (LINCS) project</a>, a <a href="https://commonfund.nih.gov/">NIH Common Fund</a> program. LINCS aims to systematically profile the molecular and phenotypical outcomes of agent perturbed human cells. The variety of agents includes chemical compounds, different micro-environments, endogenous ligands, gene knockdown and overexpression. The L1000 dataset comprises of over a million gene expression profiles of chemically or genetically perturbed human cell-lines. The L1000 technology developed at the Broad Institute measures only ~1000 genes in each sample and estimates the expression of the rest of the transcriptome using an empirical model. In this way, the technology significantly reduces the cost and makes large-scale gene expression profiling possible. L1000CDS² currently includes a subset of the chemically perturbed gene expression profiles, specifically the profiles in the LJP, CPC and CPD peturbagen groups. After averaging replicates and removing insignificant signatures, we remained with 33,197 significant signatures served by the default search.These signatures cover 62 unique cell-lines and 3,924 different small molecues. The top six cell-lines with the most number of small molecule perturbations are:</p>
            <table>
                <thead>
                    <tr>
                        <th>Cell-line</th>
                        <th>Small Molecule Count</th>
                    </tr>
                </thead>
                <tbody>
                    <tr><td>MCF7</td><td>2,052</td></tr>
                    <tr><td>VCAP</td><td>1,893</td></tr>
                    <tr><td>PC3</td><td>1,803</td></tr>
                    <tr><td>HA1E</td><td>1,505</td></tr>
                    <tr><td>A375</td><td>1,499</td></tr>
                    <tr><td>A549</td><td>1,479</td></tr>
                </tbody>
            </table>
        </section>
        <section id="download">
            <h3>Signatures Available for Download</h3>
            <p>
            A large collection of 38,9031 L1000 gene expression signatures computed using the characteristic direction signatures method, including the aforementioned signatures computed from the LJP, CPC and CPD collectionssubset can be downloaded from  <a href="http://amp.pharm.mssm.edu/public/L1000CDS_download/" target="_blank">here</a>. The link provides instructions how to install a MongoDB database and then loading the data into this database for easy access.</p>
        </section>

        <section id='tutorial'>
            <h3>Tutorial</h3>
            <h3 id="input" class="subheader">Input</h3>
            <h5 id="search">Search</h5>
            <p>The entry point into L1000CDS² is to paste up/down gene lists into the up/down gene text boxes (Fig. 1) or paste a signature (Fig. 2) into the up gene text box. A signature is a list of genes and their differential expression values separated by a comma. The search button will only become enabled when both up/down genes text boxes are filled by gene lists, or when the up gene textbox is filled with a signature. Clicking the Search button, and the information for the top 50 signatures will be displayed in a table in a new page.</p>
            <div class="col">
            <img id="geneSetFig" src='images/geneSet.png'/>
                <p class='caption'>Fig. 1 Screenshot of the input text boxes filled with up/down gene lists.</p>
            </div>
            <div class="col">
                <img id="CDFig" src='images/CD.png'/>
                <p class='caption'>Fig. 2 Screenshot of the input up gene text box filled with a signature.</p>
            </div>
                <h5 id="example">Examples and Preloaded Signatures from External Resources</h5>
                <p>Clicking the <span class="highlight2">Gene-set Example</span> button will fill in an example of up/down gene lists in the text boxes for a demo search using the gene-set method. Clicking the <span class="highlight2">Signature Example</span> button will fill in an example of a signature in the up gene text box for a demo of searching with the cosine distance method (Fig. 3).</p>
                <img id="functionsFig" src="images/functions.png"/>
            <p class="caption">Fig. 3 Screenshot of the four input form functional sections on the right: Examples and Signatures, Configuration, Metadata and Recent Searches</p>
                <p>Clicking the <span class="highlight2">EBOV Signatures</span> button will open a table with three EBOV signatures at three time points (Fig. 4). Selecting a signature by a single click and clicking the Search button to perform a cosine distance search. The selected signature will be automatically filled in the up gene text box and the associated metadata will be filled in the metadata section.
                </p>
            <img id="ebovFig" class="diseases-figure" src="images/ebov.png"/>
            <p class="caption">Fig. 4 Screenshot of the EBOV signature table</p>
                <p>
                Clicking on the <span class="highlight2">Diseases Signatures</span> button will open a table of disease names and their tissue types including the GEO ID referingreferring to the original study for the disease signature (Fig. 5). The table is searchable by disease name or tissue type and can be sorted by either column. Each row represents a differential expression signature of a disease and consists of the diferentiallydifferentially expressed genes and their expression values in the disease compared with the relevant normal tissue. These signatures were calculated from gene expression data deposited in GEO. Clicking on a row will select that disease. Clicking on the search button will fill in the signature of the selected disease in the up gene text boxes and the associated metadata in the metadata section and perform a search for small molecules at the same time.
                </p>

            <img id="fig2" class="diseases-figure" src="images/diseases.png"/>
            <p class="caption">Fig. 5 Screenshot of the disease table</p>
             <p>
                Clicking the <span class="highlight2">Ligand  Signatures</span> button will open a grid of 22 consensus ligand signatures (Fig. 6). These signatures are characteristic direction signatures computed from the LJP LINCS L1000 gene expression data using the landmark genes only. Selecting a signature by a single click, and then clicking on the Search button performs a cosine distance search on a dedicated database that contains landmark genes CD signatures (cpcd-gse70138-lm-v1.0). The default mimic/reverse setting is mimic. Users can switch the mimic/reverse slider which can be found nder the grid, to search in the reverse direction.
            </p>
             <img id="figLigand" class="diseases-figure" src="images/ligand.png"/>
            <p class="caption">Fig. 6 Screenshot of the ligand table</p>

            <p>
                Clicking the <span class="highlight2">CCLE  Signatures</span> button button will open a table of CCLE cell-lines and their tissues types. Similar to the table of disease signatures, it is searchable by cell-line or tissue type and can be sorted by either column. Each row represents a differential expression signature of a cell-line and consists of the differentially expressed genes and their expression values in the cell-line compared with the rest. They were calculated from the CCLE gene expression data. Clicking on a row will select that cell-line. Clicking on the search button will fill in the signature of the selected cell-line in the up gene text boxes and the associated metadata in the metadata section and perform a search for small molecules at the same time.
            </p>
            <h5 id="config">Configuration</h5>
            <p><span class='highlight2'>Mimic/Reverse: </span>Clicking the mimic/reverse slider can be used to switch between reverse and mimic modes (Fig. 3). In the mimic mode for the gene-set search, the input up genes are intersected with the up genes of the gene expression profiles in the L1000CDS² database, and the input down genes are intersected with the down genes for each entry in the database. In the reverse mode for the gene-set search, the input up- genes are intersected with the down- genes, and the input down- genes are intersected with the up- genes. When a cosine distance search is performed, the top 50 signatures will be those of the smallest cosine distances from the input signature in the mimic mode, or those with the largest cosine distances in the reverse mode. <strong>The default mode is reverse</strong>.</p>
            <p><span class='highlight2'>Drug combinations: </span> L1000CDS2 also provides the function to search for drug combinations. To enable this feature the user need to check the “Search for drug combinations” checkbox. When searching for combinations, L1000CDS2 compares every possible pair among the top 50 signatures and computes the potential synergy for each pair. With the gene-set search, the synergy is calculated as the combined overlap of the DE genes of two signatures with the input DE genes. In a cosine distance search, the synergy is calculated as the orthogonality between two CD signatures. The rational for this is that if two perturbations are orthogonal, they may impart their overall effect through two independent pathways.</p>
            <p><span class='highlight2'>Including more small molecules in the signature search: </span>By default L1000CDS2 searches only the significant collection of signatures as determined by the characteristic direction method. Specifically we caculate a p-value for each signature by comparing the average cosine distance of its replicates to a empirical null distribution and take those with a p-value less than 0.1 as the significant signatures. If the option to “include less significant signatures”  is selected, L1000CDS2 will search all signatures irrespective of their significance. This option allows users to increase their search breadth but run the risk of detecting and prioritizing small molecules that are less likely to produce the desire effect on global expression.</p>
            <p><span class='highlight2'>Share: </span>
            Users can share their input signatures and metadata so other users can query the signatures and gene sets they submitted. To make submitted input gene sets and signatures “public” for research purposes, a checkbox is available (Fig. 3). The default is set to “No” such that users’ input is made private. Users can still share their input lists, signatures, metadata and results using the share icon on the result page (Fig. 7). Clicking on the share icon produced a permanent URL that can be shared through e-mail, publications and other documentation.
            </p>
            <h5 id="metadata">Metadata</h5>
            <p>
            Any metadata associated with the input signature can be entered in the metadata section. By default, the section provides four input fields for metadata: Tag, Cell, Perturbation and Time Point (Fig. 3). Users can add new input fields for additional types of metadata by clicking the plus sign at the bottom or remove one by clicking the minus sign on the right of each row. The minus sign will only appear when the mouse cursor is hovering over a row. The tag field is used to enter few words which are most descriptive of the input signature.
            </p>
            <h5 id="recent">Recent Researches</h5>
            <p>Recent search history will be displayed in this section as links (Fig. 3). Clicking a link will show the results for that search. Recent searches are stored in the browser's local storage buffer. Clearing browsing data would result in a loss of these records. A maximum of 50 recent searches are stored for each user’s browser.</p>
            <h3 id="result" class="subheader">Result</h3>
            <h5 id="table">Table</h5>
            <p>The search results are rendered as a paginated table with 12 entries per page (Fig. 7). Each entry provides 7 pieces of information about the signature: rank, score, perturbation, cell-line, dose, time point and overlap with the input.</p>
            <p><span class="highlight2">Rank</span>: The rank of a signature based on its score. </p>
            <p><span id="score" class="highlight2">Search score</span>: For gene-set search, The search score is the overlap between the input DE genes and the signature DE genes divided by the effective input. The effective input is the length of the intersection between the input genes and the L1000 genes since some input lists contain genes that are not present in the L1000 dataset. This includes all ~22,000 L1000 genes, not just the measured ~1000.<br>
            For the cosine distance search, the score is the cosine distance between the input signature and the L1000 characteristic direction signatures.The consensus signatures are sorted by their scores in descending order in reverse mode and in ascending order in mimic mode.</p>

            <p><span class="highlight2">Perturbation</span>:
               The perturbation column shows the names of the chemical perturbations. User can click on the three <span class="pert-link life">L</span> <span class="pert-link pubchem">P</span> <span class="pert-link drugbank">D</span> icons to look up perturbations in <a target="_blank" href="http://life.ccs.miami.edu/life/">LIFE</a>, <a href="https://pubchem.ncbi.nlm.nih.gov/" target="_blank">PubChem</a> and <a href="http://www.drugbank.ca/" target="_blank">DrugBank</a> that catalog detailed information about the chemical compound. Not every perturbation is available in all the three resources. There is also a signature column in the table.</p>
            <p><span class="highlight2">Cell-line, Dose and Time</span>:
            The cell-line, dose and time point used for generating the signature.</p>

            <img id="tableFig" src='images/table.png'/>

            <p class='caption'>Fig. 7 Screenshot of the paginated results table and the metadata header</p>
            <p><span class="highlight2">Overlap</span>:  Clicking the overlap button will show the overlapping genes (and their values) in two text boxes (Fig. 8). <span id="geneSetOverlap">If</span> the user input type is up/down gene lists, the first box will show the overlap genes between the input up genes and the signature up (down) genes and the second will show the overlap between the input down and the signature down (up) in mimic (reverse) mode. <span id="CDOverlap">If</span> the input is a signature, the first box will show genes with a positive value in input and their values in the signature; the second box will show genes with a negative value in input and their values in the signature. The signature values and input values in both boxes are expected to be mostly in the same sign in mimic mode and in the opposite sign in reverse mode. The Enrichr button under each text box will send the genes to Enrichr for enrichment analysis.</p>
            <img id="overlapFig" src="images/overlap.png"/>
            <p class='caption'>Fig. 8 Screenshot of the overlap between the input signature and a small molecule signature and an example of the target prediction table.</p>
            <p><span class="highlight2">Target prediction</span>:  L1000CDS2 contain a feature that is used to predict the target of small-molecules and drugs. This feature uses an independent external collection of gene expression studies of single gene perturbation signatures. The goal is to predict potential targets for all the small molecules profiled in the LINCS L1000 data. The single gene perturbation signatures are manually extracted from the Gene Expression Omnibus (GEO) through a crowdsourcing project. Students from the Coursera Massive Online Open Course (MOOC) <a href="https://www.coursera.org/learn/network-biology">Network Analysis in Systems Biology</a> (NASB) were asked to participate in a crowdsourcing project where they had to identify studies that perturbed a single gene in mammalian cells and where gene expression was measured before and after such perturbation. In total, 2,476 GEO signatures covering 913 gene targets were generated and are used by L1000CDS2 for target prediction. The GEO signatures consist of perturbations that both up-regulate or down-regulate the target gene. To predict targets, we compare all the L1000 characteristic direction (CD) signatures to all the GEO single gene perturbation studies signatures. We aim to find the GEO signatures that are most similar to the L1000 CD signatures, assuming that the small molecule from the L1000 collection  target the genes from the GEO studies. The top 20 GEO signatures ranked by cosine distance are listed as potential targets. Clicking on the target icon in each row will reveal the predicted targets as a table (Fig. 8). The table displays the cosine distance, the perturbed gene, the direction of regulation and the GEO ID from which the signature was extracted. Clicking on the gene symbol displays detailed information about that gene as listed on the <a href="http://amp.pharm.mssm.edu/Harmonizome/">Harmonizome web portal</a> we independently developed. Clicking on the GEO ID, the user is directed  to the GEO page from which the gene expression of the single gene perturbation was extracted.</p>
            <p><span class="highlight2">Download</span>:
            Clicking the download button in that column will download all the information about a signature as a JavaScript Object Notation (JSON) file. </p>
            <h5 id="header">Header</h5>
            <p>
           On top of the table is a header bar that provides various functions (Fig. 7):</p>
           <p><span class="highlight2">Reanalyze: </span>Clicking this button redirects the user back to the input page with input lists or signatures preloaded in the input textboxes. Users can then reanalyze their input using different configurations, or modify the associated metadata. This function also has a bearing on sharing results with others. It provides a way for users to reanalyze their input with different settings and obtain a permanent URL for each analysis.</p>
           <p><span class="highlight2">Tag: </span>This button displays the tag and search mode. Clicking on the button shows the input metadata.</p>
           <p><span class="highlight2">Diamond: </span>By clicking on this button, L1000CDS2 performs enrichment analysis on the substructures of the top ranked small-molecules. Refer to the Substructure Enrichment section for more detail about understanding the results from this feature.</p>
          <p><span class="highlight2">Cloud download: </span>Clicking on this button downloads the table as a .csv file.</p>
          <h5 id="combinations">Table of combinations</h5>
          <p>If the user chooses to search for drug combinations, a table of signature combinations will appear below the single perturbation result table. This table is also a paginated table with 14 entries per page (Fig 9). Each entry provides three pieces of information about the identified combinations: rank, synergy score and combinations. The synergy score has been described in the configuration section. The rank is based on the synergy score. The number before each chemical perturbation in the combinations column is the rank of that perturbation in the single signature result table. Clicking on a perturbation will highlight that perturbation in the single signature results table so the user can learn more detail about that perturbation. Clicking on the cloud download button on the upper right corner downloads the combination table as a .csv file.</p>
          <img id="combinations" src="images/combinations.png"/>
          <p class='caption'>Fig. 9 Screenshot of the drug combination table results</p>
          <h5 id="substructures">Substructure Enrichment Analysis</h5>
          <p>The results of the substructure enrichment analysis are displayed as a table where each row is a significantly enriched substructure (Fig. 10). Each row provides three pieces of information: substructure, p-value and perturbation count. The substructure is represented as a string in the <a target="_blank" href="http://www.daylight.com/dayhtml/doc/theory/theory.smarts.html">SMiles ARbitrary Target Specification</a> (SMARTS) format. The p-value is computed using the Fisher’s exact test. The perturbation count shows the number of perturbations that have this substructure. Clicking on the share icon produces a permanent URL that can be share the substructure enrichment analysis results through e-mail, publications and other documentation. </p>
          <img id="combinations" src="images/substructures.png"/>
          <p class='caption'>Fig. 10 Screenshot of the table of significantly enriched substructures</p>
          <p>Clicking on the plus sign shows a visualization of the substructure and a table of the top perturbations that contain the substructure (Fig. 11). The rank in the table is the rank of the perturbations in the top 50 signature table. </p>
           <img id="combinations" src="images/substructure.png"/>
          <p class='caption'>Fig.11 Screenshot of an expanded row</p>
        </section>

        <section id='api'>
            <h3>API</h3>
            <p><strong>[POST] http://amp.pharm.mssm.edu/L1000CDS2/query</strong></p>
            <h5 id="geneSet">Gene-set Search</h5>
            <table>
                <caption>Payload <span> (content-type: application/json)</span></caption>
                <tr>
                    <td>data</td>
                    <td>Object</td>
                    <td>An object that saves the input up/down gene lists.</td>
                </tr>
                <tr>
                    <td>data.upGenes</td>
                    <td>[String]</td>
                    <td>An array of up-regulated genes.</td>
                </tr>
                <tr>
                    <td>data.dnGenes</td>
                    <td>[String]</td>
                    <td>An array of down-regulated genes.</td>
                </tr>
                 <tr>
                    <td>config</td>
                    <td>Object</td>
                    <td>An object that saves the search configuration.</td>
                </tr>
                <tr>
                    <td>config.aggravate</td>
                    <td>Object</td>
                    <td>True to perform the search in aggravate mode and False in reverse mode.</td>
                </tr>
                <tr>
                    <td>config.seachMethod</td>
                    <td>String</td>
                    <td>“geneSet”</td>
                </tr>
                <tr>
                    <td>config.share</td>
                    <td>Boolean</td>
                    <td>True to agree to share input data and metadata.</td>
                </tr>
                <tr>
                    <td>config.combination</td>
                    <td>Boolean</td>
                    <td>True to search for drug combinations.</td>
                </tr>
                 <tr>
                    <td>config['db-version']</td>
                    <td>String</td>
                    <td>The database version. Currently there are three options: 'latest', 'cpcd-gse70138-v1.0' and 'cpcd-v1.0'. The 'latest' will always point to the most recent db version. The most recent db version is 'cpcd-gse70138-v1.0'.</td>
                </tr>
                <tr>
                    <td>metadata</td>
                    <td>[Object]</td>
                    <td>An array of objects that saves the metadata of the input.</td>
                </tr>
                <tr>
                    <td>metadata[ ].key</td>
                    <td>String</td>
                    <td>A metadata field.</td>
                </tr>
                <tr>
                    <td>metadata[ ].value</td>
                    <td>String</td>
                    <td>A metadata value.</td>
                </tr>
            </table>
            <table>
                <caption>Response <span>(JSON)</span></caption>
                <tr>
                    <td>shareId</td>
                    <td>[Object]</td>
                    <td>Unique ID for sharing the search results</td>
                </tr>
                <tr>
                    <td>combinations</td>
                    <td>[Object]</td>
                    <td>Drug combinations. Only available if config.combination is set to true.</td>
                </tr>
                <tr>
                    <td>combinations[ ].X1</td>
                    <td>String</td>
                    <td>Unique identifier for a signature in the combination (sig_id).</td>
                </tr>
                <tr>
                    <td>combinations[ ].X2</td>
                    <td>String</td>
                    <td>Unique identifier for a signature in the combination (sig_id).</td>
                </tr>
                <tr>
                    <td>combinations[ ].value</td>
                    <td>Number</td>
                    <td>Synergy score</td>
                </tr>
                <tr>
                    <td>topMeta</td>
                    <td>[Object]</td>
                    <td>Descriptive information of the top 50 consensus signatures</td>
                </tr>
                <tr>
                    <td>topMeta[ ].score</td>
                    <td>Double</td>
                    <td>Score</td>
                </tr>
                <tr>
                    <td>topMeta[ ].cell_id</td>
                    <td>String</td>
                    <td>Cell-line</td>
                </tr>
                <tr>
                    <td>topMeta[ ].pert_desc</td>
                    <td>String</td>
                    <td>Perturbation name</td>
                </tr>
                <tr>
                    <td>topMeta[ ].pert_id</td>
                    <td>String</td>
                    <td>Unique identifier for a perturbation</td>
                </tr>
                <tr>
                    <td>topMeta[ ].pubchem_id</td>
                    <td>String</td>
                    <td>PubChem ID of the perturbation if exsiting</td>
                </tr>
                <tr>
                    <td>topMeta[ ].drugchem_id</td>
                    <td>String</td>
                    <td>DrugBank ID of the perturbation if existing</td>
                </tr>
                <tr>
                    <td>topMeta[ ].pert_time</td>
                    <td>String</td>
                    <td>Time point</td>
                </tr>
                <tr>
                    <td>topMeta[ ].pert_time_unit</td>
                    <td>String</td>
                    <td>Time point unit</td>
                </tr>
                <tr>
                    <td>topMeta[ ].pert_dose</td>
                    <td>String</td>
                    <td>Dose</td>
                </tr>
                 <tr>
                    <td>topMeta[ ].pert_dose_unit</td>
                    <td>String</td>
                    <td>Dose unit</td>
                </tr>
                 <tr>
                    <td>topMeta[ ].sig_id</td>
                    <td>String</td>
                    <td>Unique identifier for a signature</td>
                </tr>
                <tr>
                    <td>topMeta[ ].overlap</td>
                    <td>Object</td>
                    <td>The overlapping genes between input genes and signature DE genes.</td>
                </tr>
                <tr>
                    <td>topMeta[ ].overlap.up/dn (reverse mode only)</td>
                    <td>[String]</td>
                    <td>The overlap between input up genes and signature down genes.</td>
                </tr>
                <tr>
                    <td>topMeta[ ].overlap.dn/up (reverse mode only)</td>
                    <td>[String]</td>
                    <td>The overlap between input down genes and signature up genes.</td>
                </tr>
                <tr>
                    <td>topMeta[ ].overlap.up/up (mimic mode only)</td>
                    <td>[String]</td>
                    <td>The overlap between input up genes and signature up genes.</td>
                </tr>
                <tr>
                    <td>topMeta[ ].overlap.dn/dn (mimic mode only)</td>
                    <td>[String]</td>
                    <td>The overlap between input down genes and signature down genes.</td>
                </tr>
            </table>
            <h5 id="cosine">Cosine Distance Search</h5>
            <table>
                <caption>Payload <span> (content-type: application/json)</span></caption>
                <tr>
                    <td>data</td>
                    <td>Object</td>
                    <td>An object that saves the input signature.</td>
                </tr>
                <tr>
                    <td>data.genes</td>
                    <td>[String]</td>
                    <td>An array of input genes.</td>
                </tr>
                <tr>
                    <td>data.vals</td>
                    <td>[Number]</td>
                    <td>An array of input values that match to each input gene.</td>
                </tr>
                 <tr>
                    <td>config</td>
                    <td>Object</td>
                    <td>An object that saves the search configuration.</td>
                </tr>
                <tr>
                    <td>config.aggravate</td>
                    <td>Object</td>
                    <td>True to perform the search in aggravate mode and False in reverse mode.</td>
                </tr>
                <tr>
                    <td>config.seachMethod</td>
                    <td>String</td>
                    <td>"CD"</td>
                </tr>
                <tr>
                    <td>config.share</td>
                    <td>Boolean</td>
                    <td>True to agree to share input data and metadata.</td>
                </tr>
                <tr>
                    <td>config.combination</td>
                    <td>Boolean</td>
                    <td>True to search for drug combinations.</td>
                </tr>
                <tr>
                    <td>config['db-version']</td>
                    <td>String</td>
                    <td>The database version. Currently there are three options: 'latest', 'cpcd-gse70138-v1.0' and 'cpcd-v1.0'. The 'latest' will always point to the most recent db version. The most recent db version is 'cpcd-gse70138-v1.0'.</td>
                </tr>
                <tr>
                    <td>metadata</td>
                    <td>[Object]</td>
                    <td>An array of objects that saves the metadata of the input.</td>
                </tr>
                <tr>
                    <td>metadata[ ].key</td>
                    <td>String</td>
                    <td>A metadata field.</td>
                </tr>
                <tr>
                    <td>metadata[ ].value</td>
                    <td>String</td>
                    <td>A metadata value.</td>
                </tr>
            </table>
            <table>
                <caption>Response <span>(JSON)</span></caption>
                <tr>
                    <td>shareId</td>
                    <td>String</td>
                    <td>Unique ID for sharing the search results</td>
                </tr>
                <tr>
                    <td>uniqInput</td>
                    <td>Object</td>
                    <td>An object that stores unique input genes overlapped with L1000 genome and their averaged input values.</td>
                </tr>
                <tr>
                    <td>uniqInput.up</td>
                    <td>Object</td>
                    <td>An object that stores Unique input up genes and their averaged input values.</td>
                </tr>
                <tr>
                    <td>uniqInput.up.genes</td>
                    <td>[String]</td>
                    <td>Unique input up genes.</td>
                </tr>
                <tr>
                    <td>uniqInput.up.vals</td>
                    <td>[Number]</td>
                    <td>Values of unique input up genes.</td>
                </tr>
                <tr>
                    <td>uniqInput.dn</td>
                    <td>Object</td>
                    <td>An object that stores Unique input down genes and their averaged input values.</td>
                </tr>
                <tr>
                    <td>uniqInput.dn.genes</td>
                    <td>Object</td>
                    <td>Unique input down genes.</td>
                </tr>
                <tr>
                    <td>uniqInput.dn.vals</td>
                    <td>Object</td>
                    <td>Values of unique input down genes.</td>
                </tr>
                 <tr>
                    <td>combinations</td>
                    <td>[Object]</td>
                    <td>Drug combinations. Only available if config.combination is set to true.</td>
                </tr>
                <tr>
                    <td>combinations[ ].X1</td>
                    <td>String</td>
                    <td>Unique identifier for a signature in the combination (sig_id).</td>
                </tr>
                <tr>
                    <td>combinations[ ].X2</td>
                    <td>String</td>
                    <td>Unique identifier for a signature in the combination (sig_id).</td>
                </tr>
                <tr>
                    <td>combinations[ ].value</td>
                    <td>Number</td>
                    <td>Synergy score</td>
                </tr>
                <tr>
                    <td>topMeta</td>
                    <td>[Object]</td>
                    <td>Descriptive information of the top 50 consensus signatures</td>
                </tr>
                <tr>
                    <td>topMeta[ ].score</td>
                    <td>Double</td>
                    <td>Score</td>
                </tr>
                <tr>
                    <td>topMeta[ ].cell_id</td>
                    <td>String</td>
                    <td>Cell-line</td>
                </tr>
                <tr>
                    <td>topMeta[ ].pert_desc</td>
                    <td>String</td>
                    <td>Perturbation name</td>
                </tr>
                <tr>
                    <td>topMeta[ ].pert_id</td>
                    <td>String</td>
                    <td>Unique identifier for a perturbation</td>
                </tr>
                 <tr>
                    <td>topMeta[ ].pubchem_id</td>
                    <td>String</td>
                    <td>PubChem ID of the perturbation if exsiting</td>
                </tr>
                <tr>
                    <td>topMeta[ ].drugchem_id</td>
                    <td>String</td>
                    <td>DrugBank ID of the perturbation if existing</td>
                </tr>
                <tr>
                    <td>topMeta[ ].pert_time</td>
                    <td>String</td>
                    <td>Time point</td>
                </tr>
                <tr>
                    <td>topMeta[ ].pert_time_unit</td>
                    <td>String</td>
                    <td>Time point unit</td>
                </tr>
                <tr>
                    <td>topMeta[ ].pert_dose</td>
                    <td>String</td>
                    <td>Dose</td>
                </tr>
                 <tr>
                    <td>topMeta[ ].pert_dose_unit</td>
                    <td>String</td>
                    <td>Dose unit</td>
                </tr>
                 <tr>
                    <td>topMeta[ ].sig_id</td>
                    <td>String</td>
                    <td>Unique identifier for a signature</td>
                </tr>
                <tr>
                    <td>topMeta[ ].overlap</td>
                    <td>Object</td>
                    <td>A object that stores the signature values of unique input genes overlapped with L1000 genome.</td>
                </tr>
                <tr>
                    <td>topMeta[ ].overlap.up</td>
                    <td>[Number]</td>
                    <td>The signature values of unique input up genes.</td>
                </tr>
                <tr>
                    <td>topMeta[ ].overlap.dn</td>
                    <td>[Number]</td>
                    <td>The signature values of unique input down genes.</td>
                </tr>
            </table>
            <table id="python"><caption>P<span>ython 3</span> Example</caption></table>
            <pre>
import requests
import json
url = 'http://amp.pharm.mssm.edu/L1000CDS2/query'

def upperGenes(genes):
    # The app uses uppercase gene symbols. So it is crucial to perform upperGenes() step.
    return [gene.upper() for gene in genes]

# gene-set search example
data = {"upGenes":["KDM5A","EGR1","RELB"],
"dnGenes":["USP22","PHGDH","HADH"]}
data['upGenes'] = upperGenes(data['upGenes'])
data['dnGenes'] = upperGenes(data['dnGenes'])
config = {"aggravate":True,"searchMethod":"geneSet","share":True,"combination":True,"db-version":"latest"}
metadata = [{"key":"Tag","value":"gene-set python example"},{"key":"Cell","value":"MCF7"}]
payload = {"data":data,"config":config,"meta":metadata}
headers = {'content-type':'application/json'}
r = requests.post(url,data=json.dumps(payload),headers=headers)
resGeneSet = r.json()

# cosine distance search example
data = {"genes":["DDIT4","HIG2","FLT1","ADM","SLC2A3","ZNF331"],"vals":[9.97,10.16,7.66,17.80,20.29,15.22]}
data['genes'] = upperGenes(data['genes'])
config = {"aggravate":False,"searchMethod":"CD","share":True,"combination":True,"db-version":"latest"}
metadata = [{"key":"Tag","value":"CD python example"},{"key":"Cell","value":"VCAP"}]
payload = {"data":data,"config":config,"meta":metadata}
headers = {'content-type':'application/json'}
r = requests.post(url,data=json.dumps(payload),headers=headers)
resCD= r.json()
    </pre>
        </section>
        <section id='contact'>
            <h3>Contact</h3>
            <p>Please contact Avi Ma’ayan and Qiaonan Duan for comments, suggestions, and support: <a href='mailto:avi.maayan@mssm.edu'>avi.maayan@mssm.edu</a>, <a href='mailto:qiaonan.duan@mssm.edu'>qiaonan.duan@mssm.edu</a>.</p>
        </section>
    </div>
</body>
</html>
